File: Python/Square Fields.py
def calc (num_points, num_boxes, min_size):
    global output, boxlist
    if (num_boxes > k) or (min_size >= output):
        return
    if (num_points == n):
        output = min(output,min_size)
        return
    for i in range (num_boxes+1):
        ph = list(boxlist[i])
        boxlist[i][0] = min(boxlist[i][0], x[num_points])
        boxlist[i][1] = max(boxlist[i][1], x[num_points])
        boxlist[i][2] = min(boxlist[i][2], y[num_points])
        boxlist[i][3] = max(boxlist[i][3], y[num_points])
        calc (num_points+1, max(num_boxes, i+1), max(min_size, boxlist[i][1]-boxlist[i][0], boxlist[i][3]-boxlist[i][2]))
        boxlist[i] = ph
        

def clean ():
    global output, boxlist, x, y
    output = 1000000
    boxlist = [[65000, 0, 65000, 0] for w in range(16)]
    x = [0 for w in range (16)]
    y = [0 for w in range (16)]

T = int(raw_input())
q = 1
while (q <= T):
    clean()
    a = raw_input()
    dlist = [int(w) for w in a.split()]
    n = dlist[0]
    k = dlist[1]
    for l in range(n):
        c = raw_input()
        dlist = [int(w) for w in c.split()]
        x[l] = dlist[0]
        y[l] = dlist[1]
    calc(0,0,0)
    print "Case #" + str(q) + ": " + str(output)
    q += 1

File: C/SquareFields.c
#include <stdio.h>

#define MAX(x,y) (((x)>(y))?(x):(y)) //Ever wonder why "cock" is spelt with two C's?
#define MIN(x,y) (((x)<(y))?(x):(y)) //Damnit C.

int output, xcord[20], ycord[20], n, k, maxx[20], maxy[20], minx[20], miny[20]; //global vars

//num_point is the point number we are adding
//num_boxes is the number of squares we already have
//min_size is the current minimum size
void calc(int num_point, int num_boxes, int min_size){
    if (num_boxes > k || min_size >= output){
        //If we used too many boxes or
        //If the current size is too big
        return;
    }
    if (num_point == n){
        //If we have put all the points already
        //Update output to the minimum value
        output = MIN(output,min_size);
        return;
    }
    int j;
    for (j=0; j<=num_boxes; j++){
        int pa,pb,pc,pd;
        pa = minx[j]; //Store the values first
        pb = maxx[j]; //Cuz we will be modifying the variables later
        pc = miny[j]; //Essentially keeping a save state of variables
        pd = maxy[j];

        minx[j] = MIN(minx[j], xcord[num_point]); //Modify the variables
        maxx[j] = MAX(maxx[j], xcord[num_point]); //minx[j] is the left boundary of
        miny[j] = MIN(miny[j], ycord[num_point]); //square j.
        maxy[j] = MAX(maxy[j], ycord[num_point]); //maxx = right, miny = down, maxy = up

        calc(num_point+1, MAX(num_boxes, j+1), MAX(min_size, MAX(maxx[j]-minx[j], maxy[j]-miny[j]))); // Recursion is RIR

        minx[j] = pa; //Restore the variables when we are done with RIR
        maxx[j] = pb;
        miny[j] = pc;
        maxy[j] = pd;
    }
}

void clean(){ //Reinitialise the global variables.
    int qq;
    output = 1000000;
    for (qq=0; qq<20; qq++){
        maxx[qq] = maxy[qq] = xcord[qq] = ycord[qq] = 0;
        minx[qq] = miny[qq] = 65000;
    }
}

int main(){
    int T, i=0;
    scanf("%d", &T);
    while (i++ < T){
        clean();
        int a = 0;
        scanf ("%d %d", &n, &k);
        while (a < n){
            scanf ("%d %d", &xcord[a], &ycord[a]);
            a++;
        }
        calc(0, 0, 0); // Kinda like Dee-Pee
        printf ("Case #%d: %d\n", i, output);
    }
    return 0;
}
